### PR TITLE
Add details about libraries and applications in Wasm

### PR DESCRIPTION
**Pull request by technosophos**
This provides more details about how libraries and interfaces are stored in Bindle for WebAssembly.

Signed-off-by: Matt Butcher <matt.butcher@microsoft.com>

### CODE REVIEW

File: docs/webassembly.md [hunk 1]
```diff
@@ -18,31 +18,85 @@ In addition to the Bindle terminology (invoice, parcel, bindle, label), the foll
 
 ## Overview
 
-This document explains how the concept of beaming in the Stargazer/Tianyan platform works with the concept of bindles and parcels, and how the two can be combined to create _aggregate applications_ - applications that behave like a single program, though they are comprised of pieces that execute in separate runtimes. Sometimes these runtimes may execute on the same host, and other times the runtimes may be spread across multiple hosts.
+This document explains how we can combine nanoprocesses, Tainyan beaming, and bindles as ways to represent, store, and deploy _aggregate applications._
```
Comment (radu-matei) [inline docs/webassembly.md:21]: typo: _Tianyan_.

File: docs/webassembly.md [hunk 2]
```diff
@@ -18,31 +18,85 @@ In addition to the Bindle terminology (invoice, parcel, bindle, label), the foll
...
 
-When it comes to Bindle storage, the basic idea is that an aggregated application is stored as a single bindle with multiple parcels. The parcels contain the WebAssembly modules that compose that application. Parcels may also store additional data used by these aggregated applications.
+An aggregate application is an set of programs (modules) that together behave as a cohesive individual program, though they are comprised of separate binaries that each runs in its own runtime.
```
Comment (radu-matei) [inline docs/webassembly.md:23]: typo: _a set_

File: docs/webassembly.md [hunk 3]
```diff
@@ -18,31 +18,85 @@ In addition to the Bindle terminology (invoice, parcel, bindle, label), the foll
...
 - Once such decisions are made, the runtime executes the aggregate application
 - During execution, the runtime is responsible for delegating user interactions. (This may mean running a UI, or may mean determining what does run the UI)
 - Finally, when the aggregate application hits its stopping condition (program completes, user exits, fatal error, etc), it is the runtime's job to clean up
 
-The focus of this document is not the details of how the runtime performs these. Instead, the document focuses on how the runtime makes the decisions about how to fetch and load the constituent parts of an aggregated application. That is, this document describes how an application is described as a bindle, and how a Stargazer runtime should interpret the information in that bindle.
+This document focuses on how the runtime makes the decisions about how to fetch and load the constituent parts of an aggregated application.
+That is, this document describes how an application is described as a bindle, and how a Stargazer runtime should interpret the information in that bindle.
```
Comment (radu-matei) [inline docs/webassembly.md:54]: nit: this document describes how than application is described.

File: docs/webassembly.md [hunk 4]
```diff
@@ -18,31 +18,85 @@ In addition to the Bindle terminology (invoice, parcel, bindle, label), the foll
...
+
+*Tianyan:* Tianyan is a runtime that selects and executes modules in an aggregate application.
+It can distribute the execution of these modules over a variety of runtimes, and orchestrate their communication.
+For example, it may take a three-part application, Tianyan may choose to execute two modules locally, and the third in the cloud.
+
+*Beaming:* This term refers to the Tianyan process of sending a module or module reference to another trusted Tianyan runtime to execute the module on its behalf.
+
+*Nanoprocesses:* We used to refer to this as "tianyan" as well, but have since adopted the language used by BCA.
+We believe our usage of the term is the same as BCAs: A nanoprocess is a Wasm module that can execute on its
+own, but communicate to other Wasm modules via the component architecture (Module Linking, Type Interfaces) and WASI (IO Streams, IO Arrays).
```
Comment (radu-matei) [inline docs/webassembly.md:72]: Interface Types.

File: docs/webassembly.md [hunk 5]
```diff
@@ -18,31 +18,85 @@ In addition to the Bindle terminology (invoice, parcel, bindle, label), the foll
...
+The Stargazer model says, essentially, that Stargazer Applications are organized into Nanoprocesses.
+The developer tooling captures possible configurations of a Stargazer Application into a bindle,
+where each modules is represented as a parcel.
+A Tianyan-capable runtime reads the bindle invoice, constructs an appropriate representation of the application,
+and orchestrates its execution as one or more nanoprocesses running on one or more trusted hosts.
 
-The following example describe the aggregated application bindle, and show increasingly complicated models of applications.
+In this document, we will walk through multiple examples of how Stargazer Applications are described, structured,constructed, read, and executed.
+
+> A core principal of Bindle is that all members of a package must be referenced as parcels.
```
Comment (radu-matei) [inline docs/webassembly.md:86]: typo: _principle_

File: docs/webassembly.md [hunk 6]
```diff
@@ -207,6 +379,26 @@ The most important detail of this example is that the Bindle invoice provided su
...
+In this example, the `libalmanac.wasm` file comes from `example/libalmanac/5.1.31`.
+In example 1.a, we covered how library modules are declared.
+Here, the `origin` label points to the upstream library module so that an auditing agent has a way of tracing provenance.
+
+In the behind-the-scenes process, the `example/weather` developer fetched the `example/libalmanac` module, including it in their application.
+Upon producing the `example/weather/0.1.0` invoice, the developer included the `libalmanac.wasm` parcel into this invoice.
+It is up to tools which parcels are placed into the invoice. So, for example, tooling may also include the `libalmanac.witx` file as well.
+The important criterion is that the invoice should contain all of the parcels required to run the application.
+
+A WITX file may be valuable in some cases, and thus may be included. However, it may prover superfluous, as
```
Comment (radu-matei) [inline docs/webassembly.md:393]: typo: _prove_

**Review by radu-matei — APPROVED** [review]
LGTM, with a few typos.

**Comment by technosophos** [comment]
I screwed up the rebase, so am opening a new PR for this

**Pull request closed without merging**